[MIPS] Add MIPS i6400 and i6500 processors

mgoudar opened the pull request:
The i6400 and i6500 are high performance multi-core microprocessors from MIPS that provide best in class power efficiency for use in system-on-chip  (SoC) applications. i6400 and i6500 implements Release 6 of the MIPS64 Instruction Set Architecture with full hardware multithreading and hardware virtualization support.

Scheduling model shall be added in separate commit/PR.

mgoudar committed:
[MIPS] Add MIPS i6400 and i6500 processors

The i6400 and i6500 are high performance multi-core microprocessors
from MIPS that provide best in class power efficiency for use in
system-on-chip  (SoC) applications. i6400 and i6500 implements
Release 6 of the MIPS64 Instruction Set Architecture with full
hardware multithreading and hardware virtualization support.

Scheduling model shall be added in separate commit/PR.

djtodoro commented on clang/lib/Basic/Targets/Mips.cpp:
```diff
@@ -47,14 +47,16 @@ bool MipsTargetInfo::processorSupportsGPR64() const {
     {"mips32"}, {"mips32r2"}, {"mips32r3"}, {"mips32r5"}, {"mips32r6"},
     {"mips64"}, {"mips64r2"}, {"mips64r3"}, {"mips64r5"}, {"mips64r6"},
-    {"octeon"}, {"octeon+"}, {"p5600"}};
+    {"octeon"}, {"octeon+"},  {"p5600"},    {"i6400"},    {"i6500"}};
```
please clang-format this

djtodoro commented on clang/lib/Driver/ToolChains/Gnu.cpp:
```diff
@@ -1501,7 +1501,9 @@ bool clang::driver::findMIPSMultilibs(const Driver &D,
                       CPUName == "octeon+",
                   "-march=mips64r2", Flags);
-  addMultilibFlag(CPUName == "mips64r6", "-march=mips64r6", Flags);
+  addMultilibFlag(CPUName == "mips64r6" || CPUName == "i6400" ||
```
this as well needs clang-format

djtodoro commented on llvm/lib/Target/Mips/MipsSubtarget.h:
```diff
@@ -43,7 +43,7 @@ class MipsSubtarget : public MipsGenSubtargetInfo {
   };
 
-  enum class CPU { P5600 };
+  enum class CPU { P5600, I6400 };
```
do we need "i6500" here as well?

i6500 is just a multi-cluster version of i6400, right?

Please also attach/link documentation files for those CPUs.

mgoudar replied:
Thanks for quick review!
yes. i6500 is multi-cluster version of i6400 and both are based on MIPS64 Release 6 ISA.
Hence I kept single subtarget I6400 and reused it in Mips.td for both i6400 and i6500.
I have added comments in the changed file Mips.td and MipsSubTarget.h

djtodoro replied:
So, we do not use this in this patch at all, right? If that is the case, please leave this change for the Sched. Model PR.

mgoudar replied:
Hi, we need this change in Mips.td as we are defining 2 new cpu (i6400 and i6500) with NoSchedModel.
In the third commit, I have defined single SubTargetFeature and reused it for both i6400 and i6500 as both are based on same CPU.
Please review.

djtodoro reviewed (changes requested):
Thanks for working on this.

mgoudar committed:
[MIPS] Add comments to MipsSubTarget.h changes

I6400 and I6500 are based on MIPS64R6 ISA. I6500 is just
multi-cluster version of I6400. Hence defined single subtarget
I6400

brad0 commented on llvm/docs/ReleaseNotes.md:
```diff
@@ -101,6 +101,8 @@ Changes to the LoongArch Backend
 Changes to the MIPS Backend
 ---------------------------
 
+* `-mcpu=i6400` and `-mcpu=i6500` was added.
```
Were instead of was.

mgoudar replied:
Hi, thanks for the feedback. I have corrected it in the 3rd commit.

mgoudar committed:
[MIPS] Use single SubTargetFeature for both i6400 and i6500

i6500 is multicluster version of i6400. Both use same CPU.
Hence use same SubTargetFeature for both cpu definitions.

Updated ReleaseNotes.md

djtodoro commented on llvm/lib/Target/Mips/Mips.td:
```diff
@@ -238,13 +238,10 @@ def ImplP5600 : SubtargetFeature<"p5600", "ProcImpl",
                                  "MipsSubtarget::CPU::P5600",
                                  "The P5600 Processor", [FeatureMips32r5]>;
 
+// I6500 is multicluster version of I6400. Both are based on same CPU.
```
nit: `The I6500 is the multi-cluster version of the I6400. Both are based on the same CPU architecture.`

mgoudar replied:
updated the comments

mgoudar committed:
[MIPS] Update comments in Mips.td

djtodoro commented:
Lets wait another day or two, since someone may have additional comments. Thanks @mgoudar!

mgoudar commented:
> Lets wait another day or two, since someone may have additional comments. Thanks @mgoudar!

sure. thank you.

yingopq commented on clang/test/Driver/mips-abi.c:
```diff
@@ -121,6 +121,30 @@
+// RUN:   | FileCheck -check-prefix=MIPS-ARCH-I6500-N64 %s
+// MIPS-ARCH-I6500-N64: "-target-cpu" "i6500"
+// MIPS-ARCH-I6500-N64: "-target-abi" "n64"
+//
```
If we need add `// RUN: %clang --target=mips64-linux-gnu -### -c %s  -march=i6500 2>&1`?

mgoudar replied:
yes right. in case of mips64, default abi is n64. I have added test case for this case. thanks for catching that.

mgoudar committed:
[MIPS] Add test to validate abi for --target mips64-gnu-linux

Default abi for mips64-gnu-linux is n64

yingopq commented:
LGTM.

mgoudar commented:
thank you.
